yjit/src/backend/x86_64/mod.rs

```diff
@@ -89,19 +89,15 @@ impl Assembler
     // A special scratch register for intermediate processing.
     // Note: right now this is only used by LeaLabel because label_ref accepts
     // a closure and we don't want it to have to capture anything.
-    const SCRATCH0: X86Opnd = X86Opnd::Reg(R11_REG);
+    const SCRATCH0: X86Opnd = X86Opnd::Reg(R9_REG);
 
-    /// List of registers that can be used for stack temps.
-    pub const TEMP_REGS: [Reg; 5] = [RSI_REG, RDI_REG, R8_REG, R9_REG, R10_REG];
+    /// List of registers that can be used for Opnd::Stack.
+    pub const TEMP_REGS: [Reg; 5] = [RDI_REG, RSI_REG, RDX_REG, RCX_REG, R8_REG];
 
-    /// Get the list of registers from which we can allocate on this platform
-    pub fn get_alloc_regs() -> Vec<Reg>
-    {
-        vec![
-            RAX_REG,
-            RCX_REG,
-            RDX_REG,
-        ]
+    /// Get the list of registers that can be used for Opnd::InsnOut.
+    pub fn get_alloc_regs() -> Vec<Reg> {
+        // _C_ARG_OPNDS registers shouldn't be used to avoid conflict on C calls
+        vec![RAX_REG, R10_REG, R11_REG]
     }
```

Comment:
I don't want to make your life harder, but... 😅 

The problem with this is that R10 and R11 require the REX prefix to encode on x86-64, which forces all instructions using them to be one byte longer. Ideally we want to use registers with register number <= 7 for temp regs as much as possible.

Ideally, the best way to fix this is to have a little algorithm to generate the sequence of moves for method arguments. I implemented something like this in Higgs. You only need one temporary register to break cycles: https://github.com/higgsjs/Higgs/blob/1b4e5d85dd192592882fa2d30b839ed843be402f/source/jit/moves.d#L86

We could merge this as a temporary solution, but ideally we should implement a more optimal solution. What do you think?

Reply:
Thank you for the feedback. I haven't paid enough attention to REX.B with regard to the difference between R8-15 and the others. With that in mind, I agree that we should try to use as many registers without the prefix as possible.

I was trying to make my implementation simpler by compromising some edge cases (if you rule out InsnOut, you can know the exact registers in x86_split), but I'll experiment with an optimal solution first.